cranelift/codegen/src/isa/aarch64/lower.rs

```diff
@@ -142,6 +142,37 @@ pub(crate) fn input_to_shiftimm<C: LowerCtx<I = Inst>>(
     input_to_const(ctx, input).and_then(ShiftOpShiftImm::maybe_from_shift)
 }
 
+pub(crate) fn output_to_const_f128<C: LowerCtx<I = Inst>>(
+    ctx: &mut C,
+    out: InsnOutput,
+) -> Option<u128> {
+    if out.output > 0 {
+        None
+    } else {
+        let inst_data = ctx.data(out.insn);
+
+        match inst_data {
+            &InstructionData::UnaryConst {
+                opcode: _,
+                constant_handle,
+            } => {
+                let mut bytes = [0u8; 16];
+                let c = ctx
+                    .func()
+                    .dfg
+                    .constants
```

Comment:
Maybe only expose `func.dfg.constants` from `ctx`?

Reply:
Yes, agreed; there's a (possibly not adequately documented) goal of insulating the lowering code from the raw CLIF input as much as possible, to give us flexibility in the future. (E.g., we might someday use a VCode container of `InstructionData`s to carry non-SSA code directly from the wasm lowering; we'd need a little plumbing, but the trait abstraction gets us most of the way there.) Something like `ctx.get_constant_data(constant_handle)` would make sense, IMHO.